src/Paprika/Store/PageManagers/MemoryMappedPageManager.cs

```diff
@@ -80,6 +97,51 @@ public unsafe MemoryMappedPageManager(long size, byte historyDepth, string dir,
 
     protected override unsafe void* Ptr => _ptr;
 
+    public override void Prefetch(ReadOnlySpan<DbAddress> addresses)
+    {
+        var writer = _prefetches.Writer;
+
+        foreach (var address in addresses)
+        {
+            if (address.IsNull == false)
+            {
+                writer.TryWrite(address);
+            }
+        }
+    }
+
+    private async Task RunPrefetcher()
+    {
+        var reader = _prefetches.Reader;
+
+        while (await reader.WaitToReadAsync())
+        {
+            PrefetchImpl(reader, this);
```

Comment:
A basic question: Since the prefetcher runner thread is running asynchronously, is it theoretically possible that for a certain address it executes the prefetch **_after_** the actual read on that address is already done? Prefetcher will be able to find that the address is already in memory, but still these calls can effectively double the number of IOs?

Reply:
It is possible that it will run after the page is fetched. And yes, it's possible that it will be fetching over pages that are already there. We could then discuss whether to leave SSE-based soft prefetch and leave this implementation for heavier use. 

The place that I'd still see it as is, would be the `AbandonedList` that prepares the next page that will be copied to. In this case, the page is returned, then copied to etc while the next is scheduled for a prefetch. This should give it enough time. 

The next place that it works (just following the raw execution tiem) is visiting. We schedule prefetching for all the children up front and as it walks in the preorder, beside the first child the rest most likely will have time to be prefetched. Again, raw execution time shows a very nice boost (also, disk throughput is much higher)

The one place that makes it less easy to reason about is the `DataPage.TryGet`. Maybe, in this case we should use SSE prefetch. The page, if it faults, it will take more time anyway than a search across page (now ~10ns). So it might be the case that we'd like to skip issuing a costly OS call.

Reply:
Thanks for clarifying! Make sense to me to use two different variants based on the usage 😄